### ISSUE TITLE
🚀 Feature: Catalog filter by language tag, powered by linguist plugin

### ISSUE THREAD

**Issue by bforbis** [backstage/backstage#18258]
### 🔖 Feature description

Allow for the [linguist plugin](https://github.com/backstage/backstage/tree/master/plugins/linguist) to emit entity tags, allowing for users to easily search for components by language in the catalog index page.

### 🎤 Context

Often times, people ask a question like "How many services do we have written in NodeJS vs Java". We could ask for teams to self-report by manually adding tags, but automation may provide more value more quickly.

### ✌️ Possible Implementation

Linguist plugin provides a processor you can hook up to the catalog which will fetch the language breakdown of a component and then modify the entity to include the language tag if it is not already there.

Likely we would also want this configurable, for example:

- List of allowed languages to add tags for
- List of disallowed languages to not add tags for
- Number of bytes threshold for a language to exceed before the tag is added for that language

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

None

**Comment by bforbis** [comment]
I tried doing this myself, but I am just getting my feet wet with custom processors and probably don't know what I'm doing.

I tried passing all the `PluginEnvironment` into my processor and using this to try creating a connection to the linguist DB

```
private async connect(): Promise<LinguistBackendStore> {
    if (!this.linguistDbClient) {
      const dbclient = await this.env.database.getClient();
      const linguistDB = await LinguistBackendDatabase.create(dbclient);
      this.linguistDbClient = linguistDB;
    }
    return this.linguistDbClient;
  }
```

I then tried using this in preProcessEntity with the following code

```
    if (entity.kind === 'Component') {
      const entityRef = stringifyEntityRef(entity);
      try {
        const linguistDB = await this.connect();
        const { breakdown, processedDate } = await linguistDB.getEntityResults(
          entityRef,
        );
        entity.metadata.tags ||= [];
        breakdown.forEach(language => {
          if (
            language.type === 'programming' &&
            !entity.metadata.tags?.includes(language.name)
          ) {
            this.logger.info(
              `${entityRef} adding ${language.name} - ${language.type}`,
            );
            entity.metadata.tags?.push(language.name);
          }
        });
      } catch (err) {
        this.logger.warn(`Err pre-processing ${err}`);
      }
    }
    return entity;
  }
```

And this consistently threw the following error
```
Error: The migration directory is corrupt, the following files are missing: 20200511113813_init.js
```

I am going to guess I'm doing this entirely wrong and shouldn't be accessing the linguist DB directly, but I am not sure exactly what the right way to do this is. Should I be using the [LinguistAPIClient](https://github.com/backstage/backstage/blob/master/plugins/linguist-backend/src/api/LinguistBackendClient.ts) instead? This seemed much harder to build, as it requires a lot of dependencies (more than just the DB connection).

**Comment by awanlin** [comment]
TL; DR; Use the HTTP API for this 👍 

I've been thinking about how I would build this myself for some time now here's roughly my approach:

1. Create a `tags` table 
2. When processing an item in Linguist if the flag generate the tags is on then generate the tags so that they are valid based on the rules for them
3. Create a route for getting the `tags` using an `entityRef`
4. Create a `processor` that would simply fetch the tags by `entityRef` (if you don't get anything back just continue, it will be there later or eventually)
5. Add the tags to the entity
6. Add an annotation where the value is the `id` (PK from `tags` table?) for the `tags`

Reasons for some things:

- Getting linguist to generate the tags itself and having the processor do less of that work means it won't drag that down as people can have many processor and the generation can be slow as it needs to pull down all the code and scan it.
- Formatting the tags is needed as there are [specific rules](https://backstage.io/docs/features/software-catalog/descriptor-format#tags-optional) for this and I have a high confidence the languages names won't play nice with them. For example C# would need to be formatted differently somehow
- How do you handle updates? If the languages change we need a way to know the current tags that were added and then the new tags. We can't just blow away all the tags as you might have added some from other places (YAML, other providers, other processors).This is where the annotation comes in. You could do a look up using the `id` to get the existing tags and the by `entityRef` would get you the latest.

Keep in mind this is all in my head and I'm not sure the real world impact of some of these ideas but maybe this can be helpful for you

**Comment by bforbis** [comment]
> How do you handle updates?

I'm not sure how all this part works as I don't fully understand the backend.

So, doesn't the original location that generated that entity eventually re-fetch and merge back into the catalog, so shouldn't the linguist processor eventually get an entity handed to it that it has seen already but doesn't have the tags already since it is being re-generated from the location?

That makes it sounds that by just appending to the list of tags should be fine. If the language breakdown changes, eventually the entity will be refreshed by the catalog importer and the linguist processor will then omit the tag

> Create a tags table
> When processing an item in Linguist if the flag generate the tags is on then generate the tags so that they are valid based on the rules for them

Yeah, I think there isn't a need to store a new tags table here if we can just fetch the language breakdown, which is already pretty close to tags. Fetching it over the HTTP API means that it will either return no data (linguist hasn't processed yet), or it will fetch whatever was last pulled by linguist, which has its own re-processing phase configured in the router.

> TL; DR; Use the HTTP API for this 👍

Is there a best practice for how to hit the HTTP API of a plugin from a processor? Like an apiRef or something I should be passing to the processor from the catalog builder?

**Comment by awanlin** [comment]
As it stands right now the Linguist data is not stored in the entity at all, it gets pulled from its [database via its API](https://github.com/backstage/backstage/blob/6eb90068a88272601bdc030604bf436b21ae31f5/plugins/linguist-backend/src/api/LinguistBackendClient.ts#L90) to show the card on the entity's page.

We can't just append, for example we were using a CI/CD that was all based on Java but now we are using GitHub Actions which is all YAML. We've seen the Java drop off and the YAML show up in the Linguist card as we've done this change. 

When Linguist regenerates the data it does an ["upsert"](https://github.com/backstage/backstage/blob/6eb90068a88272601bdc030604bf436b21ae31f5/plugins/linguist-backend/src/db/LinguistBackendDatabase.ts#L72) so we would no longer have the previous state this is why I came up with creating the `tags` table. Maybe this needs to be changed? I'm open to that as an alternative - not doing an upsert, keeping the old data, and then using a date column to determine the latest data.

As far as I know there isn't a nice way to make these call in the Backend using the `apiRef` pattern. I've just been pulling in the `backend.baseUrl` and then building up the rest of the string as needed. Maybe @freben or @jhaals have suggestions?

Really happy to see this getting worked on, is this something you are going to contribute or is this just something you are doing internally?

**Comment by awanlin** [comment]
FYI @regicsolutions - you had asked for something like this in the PR that added the Linguist plugin so thought you might want to track this issue 👀

**Comment by bforbis** [comment]
@awanlin I tried this out and have confirmed the processor will drop the tag if the languages breakdown changes. If you have time can you try it out to see if it works for your use case?

**Comment by awanlin** [comment]
@bforbis sure, I just switched over to a new team so I might be a little slow with this. I have dedicated time for this on Friday every week so that's when I'll try it out. But I will give that PR a review now to at least get that started

**Comment by freben** [comment]
Finding the URL of a plugin to call should always be done through [the discovery API](https://github.com/backstage/backstage/blob/da1e780d936cfbb8c3b415c9cd330589f24ac05f/packages/backend-plugin-api/src/services/definitions/DiscoveryService.ts#L46) which is [in the env](https://github.com/backstage/backstage/blob/da1e780d936cfbb8c3b415c9cd330589f24ac05f/packages/backend/src/types.ts#L37).

**Comment by freben** [comment]
Yeah I think the two of you were talking about different upserts right? It's correct that the processor always runs on the raw input yaml over and over again, so it doesn't have to take into account what it emitted last time it was called.
